src/backend/distributed/commands/table.c

```diff
@@ -1185,6 +1185,26 @@ SetupExecutionModeForAlterTable(Oid relationId, AlterTableCmd *command)
 			{
 				executeSequentially = true;
 			}
+
+			/*
+			 * Postgres performs additional selects when creating constraints
```

Comment:
I think this comment is misleading because there nothing really special about partitioned tables, Postgres does this check for any fkey creation (unless created as INVALID).

And, although this change fixes the issue, I see some dark points. For example, in `PlanAlterTableStmt()` function, we're marking this constraint so that it skips the validation just to fix this issue, see the code and the comment:
```
				/*
				 * Foreign constraint validations will be done in workers. If we do not
				 * set this flag, PostgreSQL tries to do additional checking when we drop
				 * to standard_ProcessUtility. standard_ProcessUtility tries to open new
				 * connections to workers to verify foreign constraints while original
				 * transaction is in process, which causes deadlock.
				 */
				constraint->skip_validation = true;
``` 

(Well, it used to lead deadlocks, but today the error you see is expected as far as I can see).

The question is that why does Postgres perform the validation even tough we're trying to skip the validation. If you set the debug2, you'd see `DEBUG:  validating foreign key constraint "partitioning_reference_fkey"`.

And the subtle thing is that we see the error is happening on the partition table even if we create the fkey on the partitioned table (e.g., the SELECT is on `partitioning_test_2009`):
```SQL
ALTER TABLE partitioning_test ADD CONSTRAINT partitioning_reference_fkey FOREIGN KEY (id) REFERENCES partitioning_test_reference(id) ON DELETE CASCADE;
DEBUG:  validating foreign key constraint "partitioning_reference_fkey"
LOG:  issuing COPY (SELECT worker_column_1 AS id FROM (SELECT fk.id AS worker_column_1 FROM (public.partitioning_test_2009_102106 fk LEFT JOIN public.partitioning_test_reference_102138 pk ON ((pk.id OPERATOR(pg_catalog.=) fk.id))) WHERE ((pk.id IS NULL) AND (fk.id IS NOT NULL))) worker_subquery) TO STDOUT
...
```

I think it'd probably makes more sense to understand the change and its affect a bit more before merging this PR.

Reply:
We do set `skip_validation` flag before we call standard process utility. That is why validation is not kicked in for the partitioned table itself.

However, when creating restrictions on partitions, PG code does not consider that. Instead the foreign keys are read from catalog. see  where [foreign key cloning is initiated](https://github.com/postgres/postgres/blob/REL_11_STABLE/src/backend/commands/tablecmds.c#L7775) and where [FK descriptions are actually read](https://github.com/postgres/postgres/blob/REL_11_STABLE/src/backend/utils/cache/relcache.c#L4097). 

At this time, we can not merge a single PR due to travis failure unless we fix this. 

We should consider this PR as an interim fix.

PG folks still change the code regarding this. We would have to come up with a more proper fix. But the timing is not right now.

Reply:
> And, although this change fixes the issue, I see some dark points. For example, in `PlanAlterTableStmt()` function, we're marking this constraint so that it skips the validation just to fix this issue, see the code and the comment:
> 
> ```
> 				/*
> 				 * Foreign constraint validations will be done in workers. If we do not
> 				 * set this flag, PostgreSQL tries to do additional checking when we drop
> 				 * to standard_ProcessUtility. standard_ProcessUtility tries to open new
> 				 * connections to workers to verify foreign constraints while original
> 				 * transaction is in process, which causes deadlock.
> 				 */
> 				constraint->skip_validation = true;
> ```
> 
> (Well, it used to lead deadlocks, but today the error you see is expected as far as I can see).
> 
> The question is that why does Postgres perform the validation even tough we're trying to skip the validation. If you set the debug2, you'd see `DEBUG: validating foreign key constraint "partitioning_reference_fkey"`.

`NOT VALID` is not supported for partitioned tables, so setting `skip_validation` doesn't work. 

```
postgres=# CREATE TABLE ref(a int primary key);
postgres=# CREATE TABLE t1(a int) PARTITION BY LIST (a);
postgres=# CREATE TABLE t1_p1 PARTITION OF t1 FOR VALUES IN (1);
postgres=# ALTER TABLE t1 ADD CONSTRAINT t1_fk FOREIGN KEY (a) REFERENCES ref NOT VALID;

ERROR:  cannot add NOT VALID foreign key on partitioned table "t1" referencing relation "ref" 
DETAIL:  This feature is not yet supported on partitioned tables. 
```